File: Algorithms/Java/98 Validate Binary Search Tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    private TreeNode preNode = null;
    private boolean valid = true;

    public boolean isValidBST(TreeNode root) {
        this.valid = true;
        inOrder(root);
        return this.valid;
    }

    private void inOrder(TreeNode node) {
        if (node == null) {
            return;
        }

        inOrder(node.left);

        if (preNode == null) {
            preNode = node;
        }
        else {
            if (preNode.val >= node.val) {
                this.valid = false;
                return;
            }
            else {
                preNode = node;
            }
        }

        inOrder(node.right);
    }
}

File: Algorithms/Python/98 Validate Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def isValidBST(self, root: TreeNode) -> bool:
        def inorder(node):
            nonlocal preNode
            nonlocal valid
            
            if node is None:
                return
            
            inorder(node.left)

            if preNode is None:
                preNode = node
            else:
                if preNode.val >= node.val:
                    valid = False
                    return
                else:
                    preNode = node
        
            inorder(node.right)

        preNode = None
        valid = True
        inorder(root)

        return valid
